### PR TITLE
Add RazorSourceUpdater tests

### LINKED ISSUE #456: As a developer, I'm confident my changes won't break RazorSourceUpdater or its associated services

### PR DESCRIPTION
**Pull request by mjrousos**
Fixes  #456

### CODE REVIEW

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorTextReplacer.cs [hunk 1]
```diff
@@ -34,36 +34,57 @@ public RazorTextReplacer(ILogger<RazorTextReplacer> logger)
...
-                    MinimizeReplacement(ref originalText, ref updatedText);
+
+                    // Generally, it's not necessary to minimize replacements since the text should be the same in both documents
+                    // However, in the specific case of text being added before the first #line pragma in a Razor doc, it's possible
+                    // that C# unrelated to the original document will show up with the replaced text, so minimize replacements only
+                    // in the case that text is being inserted before the original first line.
+                    if (replacement.StartingLine == 0)
+                    {
+                        MinimizeReplacement(ref originalText, ref updatedText);
+                    }
```
Comment (mjrousos) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorTextReplacer.cs:80]: I'm still not entirely happy with how we minimize replacements. I'm going to re-work this but will do it in a separate PR since this is focused on implementing tests.

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContextProperties.cs [hunk 2]
```diff
@@ -1,12 +1,7 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System;
-using System.Collections;
 using System.Collections.Generic;
```
Comment (twsouthwick) [inline src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContextProperties.cs:4]: I wish an analyzer would flag these so we don't have to do it manually periodically....

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorTextReplacer.cs [hunk 3]
```diff
@@ -104,22 +129,32 @@ public void ApplyTextReplacements(IList<TextReplacement> replacements)
...
+            if (text.Any())
             {
-                // StreamSourceDoc.Lines is 0-based but line directives (as used in MappedSubText) are 1-based,
-                // so subtract one from i.
-                offset += lines[i - 1].Length + Environment.NewLine.Length;
+                yield return 0;
+            }
+
+            // Subsequent lines
+            for (var i = 0; i < text.Length; i++)
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorTextReplacer.cs:144]: I wonder if using `string.IndexOf(...)` would be more readable, and may be more efficient (I've seen some simd related implementations behind the scenes)

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/TextReplacement.cs [hunk 4]
```diff
@@ -78,8 +80,10 @@ public TextReplacement(SourceText originalText, SourceText newText, string fileP
         public override int GetHashCode()
         {
             var hashcode = default(HashCode);
+            hashcode.Add(StartingLine);
+            hashcode.Add(FilePath, StringComparer.OrdinalIgnoreCase);
             hashcode.Add(NewText.ToString(), StringComparer.Ordinal);
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/TextReplacement.cs:85]: calling `.ToString` seems like it may be a potentially heavy operation. Does NewText not implement `GetHashCode()`?

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/TextReplacement.cs [hunk 5]
```diff
@@ -78,8 +80,10 @@ public TextReplacement(SourceText originalText, SourceText newText, string fileP
...
-            hashcode.Add(OriginalText);
+            hashcode.Add(OriginalText.ToString(), StringComparer.Ordinal);
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/TextReplacement.cs:86]: Same as above

File: tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/DefaultTextMatcherTests.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+            // Act
+            var replacements = matcher.MatchOrderedSubTexts(originalTexts, newTexts);
+
+            // Assert
+            Assert.Collection(replacements, expectedReplacements.Select<TextReplacement, Action<TextReplacement>>(expected => actual => Assert.Equal(expected, actual)).ToArray());
+        }
+
+        public static IEnumerable<object[]> MatchOrderedSubTextsData =>
+            new List<object[]>
```
Comment (twsouthwick) [inline tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/DefaultTextMatcherTests.cs:53]: nit: I've often used a `yield return` syntax here. Seems like that would be slightly more readable

File: tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/MappedSubTextTestData.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
+﻿namespace Razor
```
Comment (twsouthwick) [inline tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/MappedSubTextTestData.cs:1]: I'm assuming this is generated. Can you make it clear so that we don't change it manually? Maybe comments at the top or a directory with RazorGeneratedSource or something

File: tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests.csproj [hunk 8]
```diff
@@ -10,8 +10,22 @@
...
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+
+  <ItemGroup>
+    <Compile Remove="MappedSubTextTestData.cs" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Remove="RazorUpdaterStepViews\TestViews\Simple.cshtml" />
```
Comment (twsouthwick) [inline tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests.csproj:23]: Maybe glob all cshtml to be removed

File: tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/RazorSourceUpdaterTests.cs [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+        }
+
+        [Theory]
+        [MemberData(nameof(ApplyData))]
+        public async Task ApplyTests(LocationLookup[][] diagnosticLocations, string[] expectedUpdatedFiles, TextReplacement[] expectedReplacements)
+        {
+            // Arrange
+            using var mock = GetMock("RazorUpdaterStepViews/Test.csproj", diagnosticLocations);
+            var razorDocs = await GetRazorCodeDocumentsAsync(mock).ConfigureAwait(true);
+            var context = mock.Mock<IUpgradeContext>();
```
Comment (twsouthwick) [inline tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/RazorSourceUpdaterTests.cs:183]: nit: `mock.Mock<>()` should only be used when it's for DI. If it's something you're passing into a message, it can cause problems (ie, all calls to `mock.Mock<>()` will return the same instance). Otherwise, you can use `new Mock<>()` to get a separate instance.

File: tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/RazorSourceUpdaterTests.cs [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+                else
+                {
+                    builder.RegisterInstance(Enumerable.Empty<DiagnosticAnalyzer>());
+                    builder.RegisterInstance(Enumerable.Empty<CodeFixProvider>());
+                }
+            });
+
+            var projectFile = mock.Mock<IProjectFile>();
+            var project = projectPath is not null ? mock.Mock<IProject>() : null;
+            project?.Setup(p => p.FileInfo).Returns(new FileInfo(projectPath!));
```
Comment (twsouthwick) [inline tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/RazorSourceUpdaterTests.cs:361]: this may be more readable if you just do 

```csharp
if(projectPath is not null)
{
.....
}
```

**Review by twsouthwick — APPROVED** [review]
A couple of comments, but LGTM

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContextProperties.cs (hunk 2 above)
Comment (mjrousos) [reply src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContextProperties.cs:4]: I agree!

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorTextReplacer.cs (hunk 3 above)
Comment (mjrousos) [reply src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorTextReplacer.cs:144]: Sure, I can do that. I'd say either way is similarly readable in my opinion.

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/TextReplacement.cs (hunk 4 above)
Comment (mjrousos) [reply src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/TextReplacement.cs:85]: Hm.. that's a good point. I don't think `SourceText` (which NewText is) has a GetHashCode, though, since it doesn't implement `IEquatable`.  Now that  look at the code, though, I'll just make the property a string type. We're not using any `SourceText` APIs (the type is just an artifact of where the data comes from in UA's workflow). That way, at least, we won't have multiple ToString calls.

File: tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/DefaultTextMatcherTests.cs (hunk 6 above)
Comment (mjrousos) [reply tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/DefaultTextMatcherTests.cs:53]: Good idea. I'll try that.

File: tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/MappedSubTextTestData.cs (hunk 7 above)
Comment (mjrousos) [reply tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/MappedSubTextTestData.cs:1]: Yep, good point. I'll add a comment.
